File: inst/examples/e2e-test/R/shinyInReact.R
shinyInReactUI <- function(id) {
  ns <- NS(id)
  tagList(
    h1("Shiny in React"),
    Separator(strong("Shiny component hidden in pivot")),
    Pivot(
      PivotItem(headerText = "Tab 1", Label("Hello 1")),
      PivotItem(
        headerText = "Initially hidden components",
        div(
          h4("textInput"),
          textInput(ns("textInputHidden"), NULL),
          textOutput(ns("textInputHiddenValue")),
          
          h4("textAreaInput"),
          textAreaInput(ns("textAreaInputHidden"), NULL),
          textOutput(ns("textAreaInputHiddenValue")),
          
          h4("checkboxInput"),
          checkboxInput(ns("checkboxInputHidden"), NULL),
          textOutput(ns("checkboxInputHiddenValue"))
        )
      )
    ),
    Separator(strong("Shiny components dynamic rendering")),
    uiOutput(ns("shinyInReact")),
    Separator(strong("Shiny components static rendering")),
    Stack(
      actionButton(ns("updateInputs"), "Update text", width = "100px"),
      
      h4("textInput"),
      textInput(ns("textInputStatic"), "Shiny input test"),
      textOutput(ns("textInputStaticValue")),
      
      h4("textAreaInput"),
      textAreaInput(ns("textAreaInputStatic"), NULL),
      textOutput(ns("textAreaInputStaticValue")),
      
      h4("checkboxInput"),
      checkboxInput(ns("checkboxInputStatic"), NULL),
      textOutput(ns("checkboxInputStaticValue"))
    )
  )
}

shinyInReactServer <- function(id) {
  moduleServer(id, function(input, output, session) {
    ns <- session$ns
    
    output$shinyInReact <- renderUI({
      Stack(
        h4("textInput"),
        textInput(ns("textInputDynamic"), NULL),
        textOutput(ns("textInputDynamicValue")),
        
        h4("textAreaInput"),
        textAreaInput(ns("textAreaInputDynamic"), NULL),
        textOutput(ns("textAreaInputDynamicValue")),
        
        h4("checkboxInput"),
        checkboxInput(ns("checkboxInputDynamic"), NULL),
        textOutput(ns("checkboxInputDynamicValue"))
      )
    })
    
    observeEvent(input$updateInputs, {
      updateTextInput(session, "textInputStatic", value = "new text")
    })
    
    ids <- c(
      "textInputStatic", "textAreaInputStatic", "checkboxInputStatic",
      "textInputDynamic", "textAreaInputDynamic", "checkboxInputDynamic",
      "textInputHidden", "textAreaInputHidden", "checkboxInputHidden"
    ) 
    wireInputToOutput(ids, input, output)
  })
}

File: js/cypress/integration/shinyInReact.js
describe('Hidden in pivot', () => {
  it('textInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('button[data-content="Initially hidden components"]').click().then(() => {
      cy.get('#shinyInReact-textInputHidden').type('my text').should('have.value', 'my text');

      cy.get('#shinyInReact-textInputHiddenValue').should('contain', 'Value: my text');
    });
  });

  it('textAreaInput() works, content sent to server and displayed in output', () => {
    const aText = 'some text';
    cy.visit('/');

    cy.get('button[data-content="Initially hidden components"]').click().then(() => {
      cy.get('#shinyInReact-textAreaInputHidden').type(aText).should('have.value', aText);

      cy.get('#shinyInReact-textAreaInputHiddenValue').should('contain', aText);
    });
  });

  it('checkboxInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('button[data-content="Initially hidden components"]').click().then(() => {
      cy.get('#shinyInReact-checkboxInputHidden').check().should('be.checked');

      cy.get('#shinyInReact-checkboxInputHiddenValue').should('contain', 'TRUE');
    });
  });
});

describe('Dynamic rendering', () => {
  it('textInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('#shinyInReact-textInputDynamic').type('my text').should('have.value', 'my text');

    cy.get('#shinyInReact-textInputDynamicValue').should('contain', 'my text');
  });

  it('textAreaInput() works, content sent to server and displayed in output', () => {
    const aText = 'some text';
    cy.visit('/');

    cy.get('#shinyInReact-textAreaInputDynamic').type(aText).should('have.value', aText);

    cy.get('#shinyInReact-textAreaInputDynamicValue').should('contain', aText);
  });

  it('checkboxInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('#shinyInReact-checkboxInputDynamic').check().should('be.checked');

    cy.get('#shinyInReact-checkboxInputDynamicValue').should('contain', 'TRUE');
  });
});

describe('Static rendering', () => {
  it('textInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('#shinyInReact-textInputStatic').type('my text').should('have.value', 'my text');

    cy.get('#shinyInReact-textInputStaticValue').should('contain', 'my text');
  });

  it('textInput() works when updated from server', () => {
    cy.visit('/');

    cy.get('button#shinyInReact-updateInputs').click().then(() => {
      cy.get('#shinyInReact-textInputStatic').should('have.value', 'new text');
    });
  });

  it('textAreaInput() works, content sent to server and displayed in output', () => {
    const aText = 'some text';
    cy.visit('/');

    cy.get('#shinyInReact-textAreaInputStatic').type(aText).should('have.value', aText);

    cy.get('#shinyInReact-textAreaInputStaticValue').should('contain', aText);
  });

  it('checkboxInput() works, content sent to server and displayed in output', () => {
    cy.visit('/');

    cy.get('#shinyInReact-checkboxInputStatic').check().should('be.checked');

    cy.get('#shinyInReact-checkboxInputStaticValue').should('contain', 'TRUE');
  });
});
